psl-clj/src/psl_clj/core.clj

```diff
@@ -0,0 +1,347 @@
+        num (count dbs)]
+    (doseq [d dbs] (close-db d))
+    num))
+        
+(defn default-inference
+  "Return an app for MPE inference using the configuration."
+  [model database config-bundle]
+  (LazyMPEInference. model database config-bundle))
+
+(defn ground-rules-by-name
+  "Return all ground rules from a given collection with the given name."
+  [ground-rules rule-name]
+  (for [gr ground-rules :when (= (.getName (.getRule gr)) rule-name)] gr))
+
+(defn ground-rules-by-name-sort
+  "Return a list of ground rules sorted by rule name."
+  [ground-rules]
+  (sort (fn [g1 g2]
+          (let [n1 (.getName (.getRule g1))
+                n2 (.getName (.getRule g2))]
+            (compare n1 n2)))
+        (seq ground-rules)))
+
+(defn ground-rules-names
+  "Return a list of distinct rule names associated with the given
+  collection of ground rules."
+  [ground-rules]
+  (sort (distinct (for [gr ground-rules] (.getName (.getRule gr))))))
+
+(defn ground-rules-print-summary         
+  "Print a summary of ground rules."
+  ;; Without rule name
+  ([ground-rules]
+   (doseq [gr (ground-rules-by-name-sort ground-rules)]
+     (println (str 
+               (clojure.string/join (repeat 30 "="))
+               (.getName (.getRule gr))
+               (clojure.string/join (repeat 30 "="))
+               ))
+     (if (instance? UnweightedGroundLogicalRule gr)
+       (println (str "INFE: " (.getInfeasibility gr)))
+       (println (str "INCO: " (.getIncompatibility gr))))
+     (println (str "CLAS: " (.getSimpleName (.getClass gr))))
+     (u/printlnw (str "STRI: " gr) 100)
+     (doseq [a (.getAtoms gr)]
+       (println (str "ATOM: " (.getValue a) ":" a)))))
+  ;; With rule name
+  ([ground-rules rule-name]
+   (ground-rules-print-summary
+    (ground-rules-by-name ground-rules rule-name))))
+
+(defn ground-rules-sample
+  "Return a sample of the ground rules up to a maximum size n."
+  [ground-rules n]
+  (u/random-sample-n (seq ground-rules) n))
+
+(defn ground-rules-stratified-sample
+  "Return a sample of ground rules, with up to n of each rule."
+  [ground-rules n]
+  (for [rule-name (ground-rules-names ground-rules)
+        ground-rule (ground-rules-sample
+                     (ground-rules-by-name ground-rules rule-name)
+                     n)]
+    ground-rule))
+
+(defn rules-info
+  "Return info about rules."
+  [model]
+  (for [k (.getRules model)]
+    (if (instance? UnweightedLogicalRule k)
+      {:kind "constraint" :name (.getName k)}
+      {:kind "compatibility" :name (.getName k) :weight (.getWeight (.getWeight k))})))
+
+(defn model-new
+  "Return a new PSLModel associated with the given DataStore."
+  [data-store]
+  (Model. "" data-store))
+
+(defn mpe-inference 
+  "Call mpeInference on the inference app."
+  [inference-app]
+  (log/info "inference:: ::starting")
+  (let [result (.mpeInference inference-app)]
+    (log/info "inference:: ::done")
+    result))
+
+(defn open-db
+  "Return an open PSL database."
```

Comment:
Is the Clojure interface meant to be a fully-featured interface like Groovy, or a limited (but more user-friendly) one like the CLI?

Reply:
Some additional functions or macros may be needed to make it fully Clojure-ized, especially for V2 upgrades.  Users can use the core PSL interfaces directly from Clojure, in combination with the simpler Clojure wrappers, so all capabilities of core are already available that way.